cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+//! Overlap detection for rules in ISLE.
+
+use crate::error::{Error, Result, Source, Span};
+use crate::sema::{
+    self, ConstructorKind, Rule, RuleId, Sym, Term, TermEnv, TermId, TermKind, Type, TypeEnv,
+    TypeId, VarId,
+};
+
+/// Check for overlap.
+pub fn check(tyenv: &TypeEnv, termenv: &TermEnv) -> Result<()> {
+    let env = Env::new(tyenv, termenv);
+    let mut errors = Errors::new(termenv.rules.len());
+    for term in termenv.terms.iter() {
+        // The only isle declaration that currently produces overlap is constructors whose
+        // definition is entirely in isle.
+        if !env.is_internal_constructor(term.id) {
+            continue;
+        }
+
+        check_overlap_groups(&mut errors, &env, term);
+    }
+
+    if !errors.is_empty() {
+        let mut errors = errors.report(&env);
+        return match errors.len() {
+            1 => Err(errors.pop().unwrap()),
+            _ => Err(Error::Errors(std::mem::take(&mut errors))),
+        };
+    }
+
+    Ok(())
+}
+
+/// A node in the error graph.
+struct Node {
+    /// The number of other rules this node overlaps with.
+    degree: usize,
+
+    /// `true` entries where an edge exists to the node at that index.
+    edges: Vec<bool>,
+}
+
+impl Node {
+    /// Make a new `Node` in the error graph with the given number of total nodes.
+    fn new(len: usize) -> Self {
+        let mut edges = Vec::with_capacity(len);
+        edges.resize(len, false);
+        Self { degree: 0, edges }
+    }
+
+    /// Add an edge between this node and the other node.
+    fn add_edge(&mut self, other: RuleId) {
+        if self.edges[other.0] {
+            return;
+        }
+
+        self.degree += 1;
+        self.edges[other.0] = true;
+    }
+
+    /// Remove an edge between this node and another node.
+    fn remove_edge(&mut self, other: RuleId) {
+        if !self.edges[other.0] {
+            return;
+        }
+
+        self.degree -= 1;
+        self.edges[other.0] = false;
+    }
+}
+
+/// A graph of all the rules in the isle source, with bi-directional edges between rules that are
+/// discovered to have overlap problems.
+struct Errors {
+    /// Edges between rules indicating overlap. As the edges are not directed, the edges are
+    /// normalized by ordering the rule ids.
+    nodes: Vec<Node>,
+}
+
+impl Errors {
+    /// Make a new `Errors` graph for collecting overlap information.
+    fn new(len: usize) -> Self {
+        let mut nodes = Vec::with_capacity(len);
+        nodes.resize_with(len, || Node::new(len));
+        Self { nodes }
+    }
+
+    /// True when there are no edges in the graph.
+    fn is_empty(&self) -> bool {
+        self.nodes.iter().all(|node| node.degree == 0)
+    }
+
+    /// Condense the overlap information down into individual errors.
+    fn report(&mut self, env: &Env) -> Vec<Error> {
```

Comment:
I'd suggest making `Errors::report` consume `self`, instead of taking it by reference. You don't need it any more afterward, and anyway there's nothing useful in it: all the edges have been deleted.

Reply:
Great idea!